torchao/prototype/low_bit_optim/cpu_offload.py

```diff
@@ -0,0 +1,34 @@
+import torch
+from torch.optim.optimizer import Optimizer
+
+
+class CPUOffloadOptimizer:
+    def __init__(self, base_optimizer: Optimizer) -> None:
+        self.optim = base_optimizer
+        self.param_cpu2cuda_map = dict()
+
+        # swap param in param_groups with CPU param
+        for param_group in base_optimizer.param_groups:
+            for i, p in enumerate(param_group["params"]):
+                p_cpu = p.detach().cpu().pin_memory()
+                param_group["params"][i] = p_cpu
+                self.param_cpu2cuda_map[p_cpu] = p
+
+    @torch.no_grad()
+    def step(self, closure=None):
+        # copy gradients from CUDA to CPU
+        for p_cpu, p_cuda in self.param_cpu2cuda_map.items():
+            if p_cuda.grad is not None:
+                p_cpu.grad = p_cuda.grad.to("cpu", non_blocking=True)
+                p_cuda.grad = None
+        torch.cuda.synchronize()
+
+        self.optim.step(closure)
+
+        # copy updated param from CPU to CUDA
+        for p_cpu, p_cuda in self.param_cpu2cuda_map.items():
+            p_cuda.copy_(p_cpu, non_blocking=True)
```

Comment:
For these H2D copies, the `non_blocking=True` here only means that the CPU will not be blocked. The `p_cpu` is already in pinned memory, so there is no further pinned memory consideration.

Calling `non_blocking=True` allows the CPU to proceed into the next logic whether that is logging, the next iteration data loading, etc. or whatever.

However, subsequent CUDA kernels issued in the default stream will still serialize with the H2D copies.

Reply:
I will still mention that this non_blocking is still benefiicial as it allows the cpu to enqueue all the copies and much better saturate the bw even if there is no overlap with compute.

Reply:
@albanD I wanted to understand this point better.

If you call `non_blocking=False`, then there is a `cudaDeviceSynchronize` after each copy, blocking the CPU until the copy finishes. After that, the CPU will proceed to issue the next copy, so there may be some slight gaps between each H2D copy.

The part that I am not clear on is, are you suggesting that these gaps are exactly what would hurt the overall copy bandwidth, or do you mean that if you issue back-to-back H2D Memcpys, then there is some kind of batching effect across copies that improves bandwidth? (The latter would be non-intuitive to me, so I wanted to check.)

Reply:
I guess for `non_blocking=False`, the additional `cudaDeviceSynchronize` is coupled with having to copy to paged memory as well, so that also is slower than copying to pinned memory.